system/lib/wasmfs/file_table.cpp

```diff
@@ -63,20 +75,37 @@ class StderrFile : public File {
   __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) override {
     return __WASI_ERRNO_INVAL;
   };
+
+public:
+  static std::shared_ptr<StderrFile> getSharedPtr() {
+    static const std::shared_ptr<StderrFile> stderrFile = std::make_shared<StderrFile>();
+    return stderrFile;
+  }
 };
 
 FileTable::FileTable() {
-  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StdinFile>()));
-  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StdoutFile>()));
-  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StderrFile>()));
+  entries.push_back(std::make_shared<OpenFileState>(0, StdinFile::getSharedPtr()));
+  entries.push_back(std::make_shared<OpenFileState>(0, StdoutFile::getSharedPtr()));
+  entries.push_back(std::make_shared<OpenFileState>(0, StderrFile::getSharedPtr()));
+}
+
+// Initialize default directories including dev/stdin, dev/stdout, dev/stderr.
+// Refers to same std streams in the open file table.
+RootDirectory::RootDirectory() {
+  auto devDirectory = std::make_shared<Directory>();
+  entries["dev"] = devDirectory;
+
+  devDirectory->get().setEntry("stdin", StdinFile::getSharedPtr());
+  devDirectory->get().setEntry("stdout", StdoutFile::getSharedPtr());
+  devDirectory->get().setEntry("stderr", StderrFile::getSharedPtr());
```

Comment:
It would be good to acquire the handle only once to avoid unnecessary lock traffic.